android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboardDownloaderActivity.java

```diff
@@ -113,267 +118,280 @@ public void onCreate(Bundle savedInstanceState) {
+      this.context = context;
+      this.showProgressDialog = showProgressDialog;
+    }
+
+    @Override
+    protected void onPreExecute() {
+      super.onPreExecute();
+      if (showProgressDialog) {
+        progressDialog = new ProgressDialog(context);
+        progressDialog.setMessage("Downloading keyboard...");
+        progressDialog.setCancelable(false);
+        if (!((Activity) context).isFinishing()) {
+          progressDialog.show();
+        } else {
+          cancel(true);
+          progressDialog = null;
         }
       }
+    }
 
-      @Override
-      protected Integer doInBackground(Void... voids) {
-        int ret = -1;
-        ArrayList<String> urls = new ArrayList<String>();
-
-        if (isCancelled())
-          return ret;
-
-        try {
-          String exceptionStr = "Invalid keyboard";
-          if (pkgID == null || pkgID.trim().isEmpty() ||
-            (!isCustom && (langID == null || langID.trim().isEmpty() || kbID == null || kbID.trim().isEmpty()))) {
-            throw new Exception(exceptionStr);
-          }
+    @Override
+    protected Integer doInBackground(Void... voids) {
+      int ret = -1;
 
-          String deviceType = context.getResources().getString(R.string.device_type);
-          if (deviceType.equals("AndroidTablet")) {
-            deviceType = "androidtablet";
-          } else {
-            deviceType = "androidphone";
-          }
+      if (isCancelled())
+        return ret;
 
-          // Placeholder here to attempt to process .kmp keyboard first:
+      try {
+        String exceptionStr = "Invalid keyboard";
+        if (pkgID == null || pkgID.trim().isEmpty() ||
+          (!isCustom && (langID == null || langID.trim().isEmpty() || kbID == null || kbID.trim().isEmpty()))) {
+          throw new Exception(exceptionStr);
+        }
 
-          ////
+        String deviceType = context.getResources().getString(R.string.device_type);
+        if (deviceType.equals("AndroidTablet")) {
+          deviceType = "androidtablet";
+        } else {
+          deviceType = "androidphone";
+        }
 
-          JSONParser jsonParser = new JSONParser();
-          JSONObject kbData = null;
-          String remoteUrl = "";
-          if (isCustom) {
-            if (isDirect) {
-              remoteUrl = url;
-            } else {
-              String encodedUrl = URLEncoder.encode(filename, "utf-8");
-              remoteUrl = String.format("%s%s&device=%s", kKeymanApiRemoteURL, encodedUrl, deviceType);
-            }
+        String remoteUrl = "";
+        if (isCustom) {
+          if (isDirect) {
+            remoteUrl = url;
           } else {
-            // Keyman cloud
-            remoteUrl = String.format("%slanguages/%s/%s?device=%s", kKeymanApiBaseURL, langID, kbID, deviceType);
+            String encodedUrl = URLEncoder.encode(filename, "utf-8");
+            remoteUrl = String.format("%s%s&device=%s", kKeymanApiRemoteURL, encodedUrl, deviceType);
```

Comment:
I think we decided to deprecate the non-direct version earlier. Might as well take it out now and always use direct.